AppCenter/AppCenter/Internals/Context/AuthToken/MSAuthTokenContext.m

```diff
@@ -42,22 +50,40 @@ + (void)resetSharedInstance {
   sharedInstance = nil;
 }
 
-- (NSString *)authToken {
+- (void)setAuthToken:(NSString *)authToken withAccountId:(NSString *)accountId {
+  NSArray *synchronizedDelegates;
+  BOOL isNewUser = NO;
   @synchronized(self) {
-    return _authToken;
+    self.authToken = authToken;
+    isNewUser = ![self.homeAccountId isEqualToString:accountId];
+    self.homeAccountId = accountId;
+
+    // Don't invoke the delegate while locking; it might be locking too and deadlock ourselves.
+    synchronizedDelegates = [self.delegates allObjects];
   }
+  [self invokeAuthTokenEvents:synchronizedDelegates withToken:authToken isNewUser:isNewUser];
 }
 
-- (void)setAuthToken:(NSString *)authToken {
+- (void)clearAuthToken {
   NSArray *synchronizedDelegates;
   @synchronized(self) {
-    _authToken = authToken;
+    self.authToken = nil;
+    self.homeAccountId = nil;
 
     // Don't invoke the delegate while locking; it might be locking too and deadlock ourselves.
     synchronizedDelegates = [self.delegates allObjects];
   }
-  for (id<MSAuthTokenContextDelegate> delegate in synchronizedDelegates) {
-    [delegate authTokenContext:self didReceiveAuthToken:authToken];
+  [self invokeAuthTokenEvents:synchronizedDelegates withToken:nil isNewUser:YES];
```

Comment:
Per my original comment (https://github.com/Microsoft/AppCenter-SDK-Apple/pull/1328#discussion_r260102518), I think it's confusing to call `isNewUser:YES` inside clear auth token method because it's not a new user. Due to this reason, I think we can keep them separate instead of refactoring into the `(void)invokeAuthTokenEvents:(NSArray *)delegates withToken:(NSString *)token isNewUser:(BOOL)newUser` method.

Reply:
Yeah it might not be a new user and I think we don't need to trigger the `authTokenContext:didUpdateUserWithAuthToken` if the current user is already an anonymous user. It's supposed to be cleared out already.